### ISSUE TITLE
Show what troops I can upgrade at my TH, and what it costs

### ISSUE THREAD

**Issue by raysun** [raysun/Clash-of-Clans-Notifier#47]
Per CLY

**Comment by raysun** [comment]
Can't be done until #44 is done.

**Comment by Peter-AC** [comment]
Well, the first thing to decide is whether to be accurate or helpful. Accuracy requires tying troop level max to lab level, not TH level. Helpful is to bypass that intermediary step, presuming the user is not pedantic and stupid, and just go with TH level. If you can't tell, I favor the latter.

Adding the data to the Troop, Spell, and Hero enums to know max level per TH level is not too bad, though it will severely run afoul of PEP's line width rules. And adding the processing on Player objects to enumerate TSH and determine which are maxed is also not difficult.

But suggesting the next one... now that is a problem. A really big problem. It's possibly even hairier than making good war CC recommendations for every clan.

The first problem is that level 1 of all Troops and Spells is unlocked through buildings, so going from 0 to 1 is different than every other upgrade.

The second problem is like it, and that is Hero upgrades are all through builders, not the lab. That becomes a factor in suggesting upgrades, potentially.

The third factor to consider is the player's resources. The important ones are total builder count, free builder count, current elixir store, current gold store, and current DE store.

The fourth factor is the player's base. Well, base + TH level. Just got into TH8 from a fully leveled TH7? Rushed TH9 from TH6? Current wall, barracks, AD, cannon, etc. levels are a consideration.

The fifth factor to consider is the intention, role, or type of account. Is it engineered? Is it going for a TH8.5? Is the player a donator, in which case maxing common donation types is favored. Is the player a war CC filler?

The sixth factor to consider is the player's attack strategies. Which do they have now, and which are they trying to get to. Should we favor improving an existing one, or improving another that has greater potential once maxed?

And the seventh is optional, but I might as well list it. In smaller clans, it may be advantageous for the leaders to split which troops/spells they max out to retain coverage but optimize advancement speed.

All of this is formalizable, but is it worth it? And how much will players need to input into the command to answer the questions the algorithm cannot automatically determine from the data?

Note that the full processing of this question starts to overlap #6.

**Comment by IamCLY** [comment]
This feature is not for suggestion which troops is good to update. But more to help player to see which troops need to max in player's TH.

Name EagleEye, TH9
Tag #222Q20PL8

Troops
Barbarian 5 of 6
Archer 5 of 6
Giant 4 of 6

Coz sometimes I need to know how many level I can upgrade my Lava etc.
Right now I have to check what level is my lab then check /s lava

**Comment by raysun** [comment]
I think we're all saying the same thing. There's 2 features: 

- Simple. Just look at my troop levels & show what's not max for the TH and/or Lab level. 
- Complicated & is for v3 - advice

**Comment by Peter-AC** [comment]
Showing what all you can upgrade, how many times, and at what cost is a useful feature, I think. The game offers almost that with the lab, except the Heroes are not included and it gives you no indication how many more levels are left at your TH. That is especially the case for things you have yet to unlock at your TH level.

**Comment by raysun** [comment]
Right. I can easily envision useful integrations both with a /UPGRADE summary command, and within the /SEARCH command.

With /Search and #44 , it would be extra-rockin', since it could even highlight the relevant stats.
